content/blog/2019-11-14-mempass/index.md

```diff
@@ -0,0 +1,292 @@
+
+Our testing procedure is as follows:
+1. Translate each test file to LLVM IR with `clang` and the `disable-O0-optnone` flag.
+2. Run MemPass on the new LLVM IR with the `opt` tool.
+3. Translate runtime libraries to LLVM IR with `clang` and the `O3` flag.
+4. Use `llvm-link` to link all files together
+5. Compile this final, linked file with `clang` and run it with the bash `time` tool. Log the user time measurements (not the sys or real).
+
+All tests were run on an Intel Core i7-7700HQ CPU @ 2.80GHz with 16GB of RAM, and using Ubuntu on WSL.
+
+### Benchmark Testing: Runtime Overhead
+
+In order to evaluate the overhead cost due to MemPass, we instrumented a subset of the LLVM Test Suite (specifically the Stanford Test Suite). Each benchmark was run 5 times, and the average and standard deviation were recorded in the following tables.
+
+| Program (Original)  | Average (s)  | SD (s)   |
+| -------- |   --------   | -------- |
+| BubbleSort     | 0.048     | 0.009     |
+| FloatMM        | 1.220     | 0.086     |
+| IntMM          | 0.000     | 0.000     |
+| Oscar          | 0.010     | 0.000     |
+| Perm           | 0.030     | 0.000     |
+| Queens         | 0.010     | 0.000     |
+| Quicksort      | 0.064     | 0.004     |
+| RealMM         | 0.010     | 0.000     |
+| Towers         | 0.028     | 0.009     |
+| Treesort       | 0.106     | 0.019     |
+
+| Program (MemPass)  | Average (s)  | SD (s)   |
+| -------- |   --------   | -------- |
+| BubbleSort     | 44.834    | 0.1995    |
+| FloatMM        | 166.354   | 2.174     |
+| IntMM          | 0.244     | 0.012     |
+| Oscar          | 1.270     | 0.016     |
+| Perm           | 2.322     | 0.186     |
+| Queens         | 8.198     | 0.252     |
+| Quicksort      | 19.060    | 0.240     |
+| RealMM         | 0.446     | 0.013     |
+| Towers         | 2.852     | 0.279     |
+| Treesort       | 5.948     | 0.305     |
+
+> **Note:** Some programs had an almost negligible runtime, which we record as 0.000 in the charts.
+
+MemPass's overhead can be anywhere from a 100x slowdown to over 1000x in the case of Bubblesort. This can be attributed to the fact that MemPass adds instrumentation not only after `malloc` and `free` calls, but also after every `load` and `store`. The benchmarks in question (especially FloatMM, IntMM, and RealMM) do many matrix computations, which end up blowing up the performance cost of our instrumentation significantly. In hindsight, restricting instrumentation to only `malloc` and `free` would significantly reduce the overhead of MemPass. As explained before, we could find some clever way to invalidate pointers once they are freed, thus causing the program to exit when trying to access an invalid location in memory.

```

Comment:
I don't think 100x is crazy at all for a heavyweight memory instrumentation pass. You can compare to the overheads for Valgrind for example.

You might also find that the overheads would be lower if you ran your pass after some standard `-O2` optimizations, which would reduce the number of loads & stores by keeping some things in SSA variables.